compiler/rustc_target/src/spec/mod.rs

```diff
@@ -1439,15 +1431,15 @@ impl Target {
             // These ABI kinds are ignored on non-x86 Windows targets.
             // See https://docs.microsoft.com/en-us/cpp/cpp/argument-passing-and-naming-conventions
             // and the individual pages for __stdcall et al.
+            //
+            // NB: these are already windows-only.
             Abi::Stdcall { unwind } | Abi::Thiscall { unwind } => {
-                if self.is_like_windows && self.arch != "x86" { Abi::C { unwind } } else { abi }
+                assert!(self.is_like_windows);
+                if self.arch != "x86" { Abi::C { unwind } } else { abi }
             }
             Abi::Fastcall | Abi::Vectorcall => {
-                if self.is_like_windows && self.arch != "x86" {
-                    Abi::C { unwind: false }
-                } else {
-                    abi
-                }
+                assert!(self.is_like_windows);
```

Comment:
These conventions are not attached to OS and can be used on non-Windows systems (https://github.com/rust-lang/rust/pull/86231#issuecomment-866326897), for example, GCC can generate them on Linux if necessary - https://gcc.gnu.org/onlinedocs/gcc/x86-Function-Attributes.html